### PR TITLE
Introduce StatefulController in TMaster and handle all stateful messages

### PR DESCRIPTION
**Pull request by srkukarni**
This pr instantiates Stateful Controller for Stateful topologies which handles all stateful related messages.

### CODE REVIEW

File: heron/tmaster/src/cpp/manager/tmaster.cpp [hunk 1]
```diff
@@ -461,12 +583,16 @@ void TMaster::DeActivateTopology(VCallback<proto::system::StatusCode> cb) {
...
+}
+
+void TMaster::HandleStatefulCkptSave(std::string _oldest_ckpt) {
+  ckptmgr_client_->SendCleanStatefulCheckpointRequest(_oldest_ckpt, false);
 }
 
 // Called when ckptmgr completes the clean stateful checkpoint request
 void TMaster::HandleCleanStatefulCheckpointResponse(proto::system::StatusCode _status) {
-  // TODO(nlu): to be implemented later
+  controller_->HandleCleanStatefulCheckpointResponse(_status);
```
Comment (nlu90) [inline heron/tmaster/src/cpp/manager/tmaster.cpp]: nitpick: s/controller_/tmater_controller_ to differentiate with the stateful_controller_

File: heron/tmaster/src/cpp/manager/tmaster.cpp [hunk 2]
```diff
@@ -589,11 +715,19 @@ void TMaster::SetPhysicalPlanDone(proto::system::PhysicalPlan* _pplan,
     auto cb = [this](EventLoop::Status status) { this->DoPhysicalPlan(status); };
     CHECK_GE(eventLoop_->registerTimer(std::move(cb), false, 0), 0);
   } else {
+    bool first_time_pplan = current_pplan_ == NULL;
```
Comment (nlu90) [inline heron/tmaster/src/cpp/manager/tmaster.cpp]: nullptr

File: heron/tmaster/src/cpp/manager/tmaster.h [hunk 3]
```diff
@@ -143,6 +159,9 @@ class TMaster {
 
   void UpdateProcessMetrics(EventLoop::Status);
 
+  // Function called when a new stateful ckpt record is saved
+  void HandleStatefulCkptSave(std::string _oldest_ckpt);
```
Comment (nlu90) [inline heron/tmaster/src/cpp/manager/tmaster.h]: s/Ckpt/Checkpoint

File: heron/tmaster/src/cpp/manager/tmaster.h [hunk 4]
```diff
@@ -127,6 +134,15 @@ class TMaster {
   // Function called after we get the topology
   void GetTopologyDone(proto::system::StatusCode _code);
 
+  // Function called after we get StatefulConsistentCheckpoints
+  void GetStatefulCheckpointDone(proto::ckptmgr::StatefulConsistentCheckpoints* _ckpt,
```
Comment (nlu90) [inline heron/tmaster/src/cpp/manager/tmaster.h]: s/GetStatefulCheckpointDone/GetStatefulCheckpointsDone

And the same for the following methods.

File: heron/tmaster/src/cpp/manager/tmaster.cpp (hunk 1 above)
Comment (srkukarni) [inline heron/tmaster/src/cpp/manager/tmaster.cpp]: Can I do it in the next pr? My feeling is that this name change will make this change confusing.

File: heron/tmaster/src/cpp/manager/tmaster.cpp (hunk 2 above)
Comment (srkukarni) [inline heron/tmaster/src/cpp/manager/tmaster.cpp]: fixed

File: heron/tmaster/src/cpp/manager/tmaster.h (hunk 4 above)
Comment (srkukarni) [inline heron/tmaster/src/cpp/manager/tmaster.h]: changed

File: heron/tmaster/src/cpp/manager/tmaster.h (hunk 3 above)
Comment (srkukarni) [inline heron/tmaster/src/cpp/manager/tmaster.h]: renamed

File: heron/tmaster/src/cpp/manager/tmaster.cpp (hunk 1 above)
Comment (nlu90) [inline heron/tmaster/src/cpp/manager/tmaster.cpp]: sure

**Pull request merged**